### PR TITLE
New Autodocs!

### PR DESCRIPTION
**Pull request by kristoff-it**
It's doctime, also aiming for that sweet https://github.com/ziglang/zig/issues/5 contributor spot

### CODE REVIEW

**Comment by andrewrk** [comment]
Looks like a false positive compile error which is a stage2 bug:

```
stage2/bin/zig build -p stage3 -Dstatic-llvm -Dtarget=native-native-musl --search-prefix /deps/local
/workspace/src/Autodoc.zig:3041:33: error: no field named 'pubDecls' in struct 'Autodoc.DocData.Type.Type__struct_34586'
/workspace/src/Autodoc.zig:485:9: note: struct declared here
```

Not sure if we have an issue open for this yet.

**Comment by andrewrk** [comment]
```
stage2/bin/zig build -Dtarget=arm-linux-musleabihf
/workspace/src/Autodoc.zig:1006:47: error: expected type 'usize', found 'u64'
/workspace/src/Autodoc.zig:1006:47: note: unsigned 32-bit int cannot represent all possible unsigned 64-bit values
```

alright @kristoff-it this one's on you, go fix your code to support 32 bit operating systems :stuck_out_tongue:

**Comment by kristoff-it** [comment]
@andrewrk  now it's the windows CI that is feeling nostalgic about the old cpp code

```
warning: FileNotFound: D:\a\1\s\src\stage1\dump_analysis.cpp
error: FileNotFound
```

**Comment by andrewrk** [comment]
Windows is the least of the problems - looks like actually running autodoc is crashing:

```
+ /workspace/_debug/staging/bin/zig test lib/std/std.zig --zig-lib-dir lib -femit-docs=/workspace/_release/staging/docs/std -fno-emit-bin
thread 144370 panic: reached unreachable code
???:?:?: 0x35dca5b in ??? (???)
???:?:?: 0x37b156c in ??? (???)
???:?:?: 0x376cad1 in ??? (???)
???:?:?: 0x3760657 in ??? (???)
???:?:?: 0x36eba8f in ??? (???)
???:?:?: 0x3642462 in ??? (???)
???:?:?: 0x35ea8e9 in ??? (???)
???:?:?: 0x35ea02c in ??? (???)
???:?:?: 0x9452752 in ??? (???)
Aborted (core dumped)
```

I'm looking into it now.

**Comment by andrewrk** [comment]
`openDirAbsolute` is not correct, you need to play nicely with the open dir handles

```
[nix-shell:~/dev/zig]$ build-release/stage1/bin/zig test lib/std/std.zig --zig-lib-dir lib -femit-docs=docs -fno-emit-bin
LLVM Emit Object... thread 58224 panic: reached unreachable code
/home/andy/dev/zig/lib/std/debug.zig:281:14: 0xad82d8 in std.debug.assert (zig1)
    if (!ok) unreachable; // assertion failure
             ^
/home/andy/dev/zig/lib/std/fs.zig:2358:11: 0xc9e53c in std.fs.openDirAbsolute (zig1)
    assert(path.isAbsolute(absolute_path));
          ^
/home/andy/dev/zig/src/Autodoc.zig:271:42: 0xc5acd6 in Autodoc.generateZirData (zig1)
    var docs_dir = std.fs.openDirAbsolute(docs, .{}) catch unreachable;
                                         ^
/home/andy/dev/zig/src/Compilation.zig:2295:40: 0xc4ed84 in Compilation.update (zig1)
            try autodoc.generateZirData();
                                       ^
/home/andy/dev/zig/src/main.zig:3294:20: 0xbdd67f in main.updateModule (zig1)
    try comp.update();
                   ^
/home/andy/dev/zig/src/main.zig:2983:17: 0xb388b3 in main.buildOutputType (zig1)
    updateModule(gpa, comp, hook) catch |err| switch (err) {
                ^
/home/andy/dev/zig/src/main.zig:225:31: 0xae5336 in main.mainArgs (zig1)
        return buildOutputType(gpa, arena, args, .zig_test);
                              ^
/home/andy/dev/zig/src/stage1.zig:48:24: 0xae4ac9 in main (zig1)
        stage2.mainArgs(gpa, arena, args) catch unreachable;
                       ^
Aborted (core dumped)
```

File: src/Autodoc.zig [hunk 1]
```diff
@@ -0,0 +1,3887 @@
...
+    if (self.doc_location.directory) |dir| {
+        if (dir.path) |path| {
+            log.debug("path: {s}", .{path});
+        }
+    }
+
+    var buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
+    const dir =
+        if (self.module.main_pkg.root_src_directory.path) |rp|
+        std.os.realpath(rp, &buf) catch unreachable
```
Comment (andrewrk) [inline src/Autodoc.zig:73]: don't call realpath, also fix all the `catch unreachable`

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Comment by kristoff-it** [comment]
@andrewrk I've reworked how file access happens, can you confirm this is the appropriate way?

**Pull request merged**